Track: Add missing `NOTIFY` signals for `Q_PROPERTY` declarations

Holzhaus opened the pull request:
This fixes some clazy-qproperty-without-notify warnings. Every
`Q_PROPERTY` should either have a `NOTIFY` signal or be declared as
`CONSTANT`.

See this for details:
- https://github.com/KDE/clazy/blob/master/docs/checks/README-qproperty-without-notify.md
- https://doc.qt.io/qt-5.12/properties.html

Holzhaus committed:
Track: Add missing `NOTIFY` signals for `Q_PROPERTY` declarations

This fixes some clazy-qproperty-without-notify warnings. Every
`Q_PROPERTY` should either have a `NOTIFY` signal or be declared as
`CONSTANT`.

See this for details:
- https://github.com/KDE/clazy/blob/master/docs/checks/README-qproperty-without-notify.md
- https://doc.qt.io/qt-5.12/properties.html

Holzhaus commented:
@daschuer With this, I can emit the proper signals in #3911 atomically. This should also improve QML performance with lots of decks/samplers.

daschuer commented on src/track/track.cpp:
```diff
@@ -84,6 +84,9 @@ Track::Track(
     }
+
+    connect(this, &Track::titleChanged, this, &Track::infoChanged);
+    connect(this, &Track::artistChanged, this, &Track::infoChanged);
```
I would prefer to call the additional signals explicit.
For less magic to consider in case of future refactoring.

Holzhaus replied:
You want me to add slots that only contain `emit infoChanged()` ?

daschuer replied:
No, I mean copy emit infoChanged() below emit artistChanged()

Than the call tree has one interruption less when debugging.

daschuer replied:
Than we can consider to get rid of infoChanged() in favor of changed()

Holzhaus replied:
No, the signal signature is incompatible.

uklotzde commented:
All those properties might also change when invoking one of the following functions:
- replaceMetadataFromSource()
- mergeExtraMetadataFromSource() (probably no properties affected)
- replaceRecord()
- exportMetadata()
- setAudioProperties()
- updateStreamInfoFromSource()

We could simply trigger all property changed signals in the first step to ensure consistency, especially for the first 4 methods.

uklotzde commented on src/track/track.h:
```diff
@@ -47,27 +47,34 @@ class Track : public QObject {
+    Q_PROPERTY(int timesPlayed READ getTimesPlayed NOTIFY timesPlayedChanged)
+    Q_PROPERTY(QString comment READ getComment WRITE setComment NOTIFY commentChanged)
+    Q_PROPERTY(double bpm READ getBpm NOTIFY bpmUpdated)
+    Q_PROPERTY(QString bpmFormatted READ getBpmText STORED false NOTIFY bpmUpdated)
```
The change signal has a different type than the property. Is this intended? I would expect that all change signals emit the actual property type.

Holzhaus replied:
The signals without parameters are intended:

> A NOTIFY signal is optional. If defined, it should specify one existing signal in that class that is emitted whenever the value of the property changes. NOTIFY signals for MEMBER variables must take zero or one parameter, which must be of the same type as the property.

https://doc.qt.io/qt-5.12/properties.html

The type mismatch is uninteded though :D

Holzhaus replied:
Done.

Holzhaus committed:
Track: Emit infoChanged signals directly when setting title/artist

Holzhaus committed:
Track: Add dedicated bpmTextChanged signal

Holzhaus committed:
Track: Emit duration changed signal when updating stream info

Holzhaus committed:
Track: Emit changed signals for all properties on metadata import

Holzhaus commented:
> * exportMetadata()

Does this really modify any metadata?

> * setAudioProperties()
> * updateStreamInfoFromSource()

This can only change the duration, right?

I added some commits, please have a look.

uklotzde commented:
Ok, exportMetadata() doesn't seem to modify any of those exposed properties. It might only update hidden, internal members.

Duration should be the only Q_PROPERTY that is affected by stream info changes.

Holzhaus commented:
Ok, then I think this is ready.

uklotzde commented on src/track/track.h:
```diff
@@ -47,27 +47,34 @@ class Track : public QObject {
+    Q_PROPERTY(int timesPlayed READ getTimesPlayed NOTIFY timesPlayedChanged)
+    Q_PROPERTY(QString comment READ getComment WRITE setComment NOTIFY commentChanged)
+    Q_PROPERTY(double bpm READ getBpm NOTIFY bpmUpdated)
+    Q_PROPERTY(QString bpmFormatted READ getBpmText STORED false NOTIFY bpmTextChanged)
```
How about consistently adopting the `Text` prefix for all formatted (= derived text) properties, i.e. `bpmFormatted` -> `bpmText` for consistency?

This also applies to `key` -> `keyText` / `keysUpdated` -> `keyTextUpdated`.

Even worse are the naming inconsistencies of the various duration properties. I didn't check how much code would be affected.

Managing all those updates and their dependencies is tedious and error prone.. Unfortunately, I have no idea how to adopt and apply FRP principles in Qt/QML or what idiomatic patterns they propose instead.

uklotzde commented on src/track/track.h:
```diff
@@ -363,6 +370,22 @@ class Track : public QObject {
+    void trackNumberChanged(const QString&);
+    void trackTotalChanged(const QString&);
+    void commentChanged(const QString&);
+    void bpmTextChanged();
```
Why do some signals carry the value and others not?

uklotzde replied:
Oh, even the docs don't apply a consistent approach:

https://doc.qt.io/qt-5/properties.html

uklotzde replied:
0 or 1 parameter are allowed, but if the new value is provided as a parameter than it's type must match the type of the property.

In order to keep the number of signals low let's apply the following rationale:
- For native properties without any derived properties provide the new value if possible
- For properties with derived properties use the name of the native property, reuse the signal for all derived properties, and don't provide any value.

By applying this rule we need to revert `bpmTextChanged` to `bpmChanged`.

uklotzde replied:
The `Track` object with its thread synchronization is a bastard and not really suitable for this purpose. What we actually would need is a distinction between *view model* and *domain model* classes.

Holzhaus replied:
https://github.com/mixxxdj/mixxx/pull/3924/commits/640a59d095e959302b1b52428b7891f864c09961

Holzhaus committed:
Track: Make property naming consistent and use new NOTIFY signal scheme

For native properties without any derived properties provide the new
value if possible. For properties with derived properties use the name
of the native property, reuse the signal for all derived properties, and
don't provide any value.

daschuer commented on src/track/track.cpp:
```diff
@@ -431,17 +433,34 @@ void Track::afterBeatsAndBpmUpdated(
     emit beatsUpdated();
 }
 
+void Track::emitMetadataChanged() {
```
This reads as if a signal "MetadataChanged" will be emitted. 
How about  emitChangedSignalsForAllMetadata() or something like that.

daschuer commented on src/track/track.cpp:
```diff
@@ -431,17 +433,34 @@ void Track::afterBeatsAndBpmUpdated(
+
 void Track::setMetadataSynchronized(bool metadataSynchronized) {
     QMutexLocker lock(&m_qMutex);
     if (compareAndSet(m_record.ptrMetadataSynchronized(), metadataSynchronized)) {
```
Is here a emitMetadataChanged() missing?

uklotzde replied:
No. There is no property for this flag.

daschuer reviewed (changes requested):
is a signal missing in: 
Track::setType(const QString& sType)
Track::setBitrate(int iBitrate)
Track::setURL(const QString& url)
Track::shiftCuePositionsMillis(double milliseconds)
Track::setRating (int rating)
Track::setBpmLocked(bool bpmLocked)

I am unsure if everything is correct in 
Track::updateStreamInfoFromSource()
There is either an emitMetadataChanged() missing, or an redundant markDirtyAndUnlock(&lock); call.

Holzhaus committed:
Track: Rename emitMetataChanged method to better reflect intent

Holzhaus committed:
Track: Emit missing durationChanged signal in updateStreamInfoFromSource

Holzhaus commented:
> is a signal missing in:
> Track::setType(const QString& sType)
> Track::setBitrate(int iBitrate)
> Track::setURL(const QString& url)
> Track::shiftCuePositionsMillis(double milliseconds)
> Track::setRating (int rating)
> Track::setBpmLocked(bool bpmLocked)

None of those is a Q_PROPERTY. We can add more if we need to, but this is out of scope here.

> I am unsure if everything is correct in
> Track::updateStreamInfoFromSource()
> There is either an emitMetadataChanged() missing, or an redundant markDirtyAndUnlock(&lock); call.

Good catch, added the missing durationChanged signal.

uklotzde commented:
@daschuer Ping

daschuer commented:
LGTM, thank you.
